packages/web/src/components/composer-input/ComposerInput.tsx

```diff
@@ -0,0 +1,363 @@
+      setValue(editedValue)
+      // TODO: Need to update this to move to the proper position affect link change to human text
+      // setTimeout(() => {
+      //   textarea.selectionStart = cursorPosition
+      //   textarea.selectionEnd = cursorPosition
+      // }, 0)
+    },
+    [resolveLinks, setValue]
+  )
+
+  const handleSubmit = useCallback(() => {
+    onSubmit?.(restoreLinks(value), linkEntities)
+  }, [linkEntities, onSubmit, restoreLinks, value])
+
+  // Submit when pressing enter while not holding shift
+  const handleKeyDown = useCallback(
+    (e: React.KeyboardEvent<HTMLTextAreaElement>) => {
+      if (isUserAutocompleteActive) {
+        if (e.key === ENTER_KEY || e.key === TAB_KEY) {
+          e.preventDefault()
+          // TODO: Fill in the top result
+        }
+
+        if (e.key === ESCAPE_KEY || e.key === SPACE_KEY) {
+          setIsUserAutocompleteActive(false)
+        }
+
+        if (e.key === BACKSPACE_KEY) {
+          const textarea = e.target as HTMLTextAreaElement
+          const cursorPosition = textarea.selectionStart
+          const deletedChar = textarea.value[cursorPosition - 1]
+          if (deletedChar === '@') {
+            setIsUserAutocompleteActive(false)
+          }
+        }
+
+        return
+      }
+
+      // Submit on enter
+      if (e.key === ENTER_KEY && !e.shiftKey) {
+        if (onSubmit) {
+          e.preventDefault()
+          handleSubmit()
+        }
+      }
+
+      // Start user autocomplete
+      if (e.key === AT_KEY) {
+        setIsUserAutocompleteActive(true)
+      }
+
+      // Delete any matched values with a single backspace
+      if (e.key === BACKSPACE_KEY) {
+        const textarea = e.target as HTMLTextAreaElement
+        const cursorPosition = textarea.selectionStart
+        const textBeforeCursor = textarea.value.slice(0, cursorPosition)
+        const { editValue, newCursorPosition } = handleBackspace({
+          cursorPosition,
+          textBeforeCursor
+        })
+        if (editValue) {
+          e.preventDefault()
+          setValue(editValue)
+          setTimeout(() => {
+            textarea.selectionStart = newCursorPosition
+            textarea.selectionEnd = newCursorPosition
+          }, 0)
+        }
+      }
+    },
+    [isUserAutocompleteActive, onSubmit, handleSubmit, handleBackspace]
+  )
+
+  const handleAutocomplete = useCallback(
+    (user: UserMetadata) => {
+      const cursorPosition = ref.current?.selectionStart || 0
+      const autocompleteRange = isUserAutocompleteActive
+        ? [value.slice(0, cursorPosition).lastIndexOf('@'), cursorPosition]
+        : [0, 1]
+      const mentionText = `@${user.handle}`
+
+      if (!userMentions.includes(mentionText)) {
+        setUserMentions((mentions) => [...mentions, mentionText])
+      }
+      setValue(
+        (value) =>
+          `${value.slice(0, autocompleteRange[0])}${mentionText}${value.slice(
+            autocompleteRange[1]
+          )}`
```

Comment:
nit: the inline `value.slice`s are rather hard to read, maybe splitting them up into variables would be better